tests/unit/test_req_file.py

```diff
@@ -498,6 +498,47 @@ def test_skip_regex(self, tmpdir, finder, options):
 
         assert finder.index_urls == ['Good']
 
+    def test_expand_existing_env_variables(self, tmpdir, finder):
+        template = ('https://%s:x-oauth-basic@'
+                    'github.com/user/%s/archive/master.zip')
+
+        env_vars = (
+            ('GITHUB_TOKEN', 'notarealtoken'),
+            ('DO_12_FACTOR', 'awwyeah'),
+        )
+
+        with open(tmpdir.join('req1.txt'), 'w') as fp:
+            fp.write(template % tuple(['${%s}' % k for k, _ in env_vars]))
+
+        with patch('pip.req.req_file.os.getenv') as getenv:
+            getenv.side_effect = lambda n: dict(env_vars)[n]
+
+            reqs = list(parse_requirements(tmpdir.join('req1.txt'),
+                                           finder=finder,
+                                           session=PipSession()))
+
+            assert len(reqs) == 1
+
+            expected_url = template % tuple([v for _, v in env_vars])
+            assert reqs[0].link.url == expected_url
+
+    def test_expand_missing_env_variables(self, tmpdir, finder):
+        req_url = ('https://${NON_EXISTING_VARIABLE}:$WRONG_FORMAT@'
```

Comment:
While the test is still reasonably accurate, isn't this supposed to be `https://${WRONG_FORMAT}:$NON_EXISTING_VARIABLE@` instead?

Reply:
:art: `NON_EXISTENT_VARIABLE`